src/backend/columnar/columnar_customscan.c

```diff
@@ -491,15 +563,508 @@ RelationIdGetNumberOfAttributes(Oid relationId)
+		return true;
+	}
+
+	Assert(sslot.nnumbers == 1);
+
+	float4 varCorrelation = sslot.numbers[0];
+
+	ReleaseVariableStats(varStatData);
+
+	/*
+	 * If the Var is not highly correlated, then the chunk's min/max bounds
+	 * will be nearly useless.
+	 */
+	if (Abs(varCorrelation) < ColumnarQualPushdownCorrelationThreshold)
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * ExprReferencesRelid returns true if any of the Expr's Vars refer to the
+ * given relid; false otherwise.
+ */
+static bool
+ExprReferencesRelid(Expr *expr, Index relid)
+{
+	List *exprVars = pull_var_clause(
+		(Node *) expr, PVC_RECURSE_AGGREGATES |
+		PVC_RECURSE_WINDOWFUNCS | PVC_RECURSE_PLACEHOLDERS);
+	ListCell *lc;
+	foreach(lc, exprVars)
+	{
+		Var *var = (Var *) lfirst(lc);
+		if (var->varno == relid)
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * CheckPushdownClause tests to see if clause is a candidate for pushing down
+ * into the given rel (including join clauses). This test may not be exact in
+ * all cases; it's used to reduce the search space for parameterization.
+ *
+ * Note that we don't try to handle cases like "Var + ExtParam = 3". That
+ * would require going through eval_const_expression after parameter binding,
+ * and that doesn't seem worth the effort. Here we just look for "Var op Expr"
+ * or "Expr op Var", where Var references rel and Expr references other rels
+ * (or no rels at all).
+ */
+static bool
+CheckPushdownClause(PlannerInfo *root, RelOptInfo *rel, Expr *clause)
+{
+	if (!IsA(clause, OpExpr) || list_length(((OpExpr *) clause)->args) != 2)
+	{
+		ereport(ColumnarPlannerDebugLevel,
+				(errmsg("columnar planner: cannot push down clause: "
+						"must be binary operator expression")));
+		return false;
+	}
+
+	OpExpr *opExpr = castNode(OpExpr, clause);
+	Expr *lhs = list_nth(opExpr->args, 0);
+	Expr *rhs = list_nth(opExpr->args, 1);
+
+	Var *varSide;
+	Expr *exprSide;
+
+	if (IsA(lhs, Var) && ((Var *) lhs)->varno == rel->relid &&
+		!ExprReferencesRelid((Expr *) rhs, rel->relid))
+	{
+		varSide = castNode(Var, lhs);
+		exprSide = rhs;
+	}
+	else if (IsA(rhs, Var) && ((Var *) rhs)->varno == rel->relid &&
+			 !ExprReferencesRelid((Expr *) lhs, rel->relid))
+	{
+		varSide = castNode(Var, rhs);
+		exprSide = lhs;
+	}
+	else
+	{
+		ereport(ColumnarPlannerDebugLevel,
+				(errmsg("columnar planner: cannot push down clause: "
+						"must match 'Var <op> Expr' or 'Expr <op> Var'"),
+				 errhint("Var must only reference this rel, "
+						 "and Expr must not reference this rel")));
+		return false;
+	}
+
+	if (varSide->varattno <= 0)
+	{
+		ereport(ColumnarPlannerDebugLevel,
+				(errmsg("columnar planner: cannot push down clause: "
+						"var is whole-row reference")));
+		return false;
+	}
+
+	if (contain_volatile_functions((Node *) exprSide))
+	{
+		ereport(ColumnarPlannerDebugLevel,
+				(errmsg("columnar planner: cannot push down clause: "
+						"expr contains volatile functions")));
+		return false;
+	}
```

Comment:
similar to https://github.com/citusdata/citus/pull/5172/files#r701265217, what do you think ?

Reply:
I tried to deparse the clause. Almost got that working with some hacks, but gave up. It seems surprisingly annoying to deparse at path generation time.